README.md
# TJ Feeder

This package generates a daily input for [TaskJuggler](https://taskjuggler.org/) from a CSV file (comma as separator) with headers: time_spent (XYmin or X.Yh), issue_name (name of task), and issue_description (optional description of task).

# Installation

The project is available on PyPI's page [TJ_Feeder](https://pypi.org/project/tj-feeder/). To install it, run:

```shell
pip install tj_feeder
```

# Set up

First, you need to set up the starting hour of your shift, the shift duration in hours, and the display mode for task durations (minutes or hours).

You will need a holidays file containing specifying holiday dates in the format **yyyy-mm-dd** (e.g. "2021-25-12"). Make sure to have only one date per line. To set the path to this file, run the command:
```shell
tj_feed define --holidays-file [path_to_file]
```

To set the starting hour of your shift, run the command:
```shell
tj_feed define --starting-hour [integer]
```

To set the shift duration, run the command:
```shell
tj_feed define --shift-hours [integer]
```

To set month starting workday ranging from 1 to 31 (i.e. the first workday accounted for the month's invoice), run the command:

```shell
tj_feed define --month-start-workday [integer]
```

To set the display mode for minutes (in case you want the display mode for hours, you can set this to False), run the command:
```shell
tj_feed define --use-minutes [True|False]
```

# Generating Daily Feed

Create your CSV file following the format **yyyy-mm-dd.csv** (e.g. 2021-09-30.csv) and fill it like this:

```
time_spent,issue_name,issue_description
<time_spent_in_minutes_or_hours>,<issue1_name>,<issue1_description>
<time_spent_in_minutes_or_hours>,<issue2_name>,<issue2_description>
<time_spent_in_minutes_or_hours>,<issue3_name>,<issue3_description>
```

- Make sure to use the same headers;
- The time spent can be in two formats: XYmin or X.Yh (e.g. 45min or 0.75h);
- Make sure the issue names match with the tasks defined by your team manager;
- The issue description is optional;
- Use commas only to separate the values.

E.g.:
```
time_spent,issue_name,issue_description
30min,communication_9,Meetings
0.5h,management_6,Weekly review
7.0h,my_proj_13,Module implementation
```
Once you have your csv file ready. You can run the following command:
```shell
tj_feed feed <your_file.csv>
```

E.g.

```shell
tj_feed feed ../test/2021-09-30.csv
booking communication_9                2021-09-10-09:00 +0.5h                # Meetings
booking management_6                   2021-09-10-09:30 +0.5h                # Weekly review
booking my_proj_13                     2021-09-10-10:00 +7.0h                # Module implementation
```

In case, the time spent is higher than the shift duration the overtime flag will be set automatically.


E.g. (assuming a shift of 8 hours long)

```shell
tj_feed feed ../test/2021-09-30.csv
booking communication_9                2021-09-30-09:00 +0.5h                # Meetings
booking management_6                   2021-09-30-09:30 +0.5h                # Weekly review
booking my_proj_13                     2021-09-30-10:00 +7.5h   {overtime 1} # Module implementation
```

# Batch processing

It's also possible to create and feed multiple CSV files in one go.

## Creating all month's CSV files

To create all necessary CSV files with headers and considering holidays (see: [Set up](#Set-up)) and weekends, run the following command:

```shell
tj_feed create_month_dir --root-directory <path to directory> --year <integer> --month <integer>
```

E.g. running this command "`tj_feed create_month_dir --root-directory my_bookings/ --year 2022 --month 1`", will result in the following structure:
```
root_directory/
|__ 2022-01/
    |__ 2022-01-03.csv
    |__ 2022-01-04.csv
    ...
    |__ 2022-01-31.csv
```

## Feed all month's CSV files

To generate a feed using the month directory created by
"`tj_feed create_month_dir`" command, you can use the
"`tj_feed feed_month_dir`" command.

```shell
tj_feed feed_month_dir --month-directory <path to directory>
```

The feed will have one extra line break to separate bookings per day
and three extra line breaks to separate bookings per week.

E.g. running the command "`tj_feed feed_month_dir --month-directory my_bookings/2022-01`", will have an output similar to this:

```
{bookings for 2022-01-03}

{bookings for 2022-01-04}

{bookings for 2022-01-05}

{bookings for 2022-01-06}

{bookings for 2022-01-07}



{bookings for 2022-01-10}

{bookings for 2022-01-11}

{bookings for 2022-01-12}

...
```


CHANGELOG.md
# Changelog

## [v0.1.0](https://github.com/MoraesCaio/TJ_Feeder/tree/v0.1.0) (2021-09-26)

[Full Changelog](https://github.com/MoraesCaio/TJ_Feeder/compare/v0.0.5...v0.1.0)

**Implemented enhancements:**

- Batch parsing process [\#5](https://github.com/MoraesCaio/TJ_Feeder/issues/5)
- Batch creation process [\#4](https://github.com/MoraesCaio/TJ_Feeder/issues/4)
- Feature/batch parsing process [\#7](https://github.com/MoraesCaio/TJ_Feeder/pull/7)
- Added batch.py with create\_month\_csv\_dir function [\#6](https://github.com/MoraesCaio/TJ_Feeder/pull/6)

**Fixed bugs:**

- Add batch commands to binary CLI [\#8](https://github.com/MoraesCaio/TJ_Feeder/issues/8)
- time\_helper.Dates.get\_month\_workdays is returning weekends too [\#2](https://github.com/MoraesCaio/TJ_Feeder/issues/2)
- Fix/batch commands cli [\#11](https://github.com/MoraesCaio/TJ_Feeder/pull/11)
- FIX: Removing month weekends from workdays list [\#3](https://github.com/MoraesCaio/TJ_Feeder/pull/3)
- Raising ValueError exceptions for workday parsing [\#1](https://github.com/MoraesCaio/TJ_Feeder/pull/1)

**Closed issues:**

- Update documentation for CLI [\#12](https://github.com/MoraesCaio/TJ_Feeder/issues/12)

**Merged pull requests:**

- Updated documentation [\#14](https://github.com/MoraesCaio/TJ_Feeder/pull/14)

## [v0.0.5](https://github.com/MoraesCaio/TJ_Feeder/tree/v0.0.5) (2021-09-18)

[Full Changelog](https://github.com/MoraesCaio/TJ_Feeder/compare/4521972e22abd011e89577364ab1d76b1bf27d96...v0.0.5)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


tj_feeder/batch.py
from datetime import datetime
from pathlib import Path
from glob import glob
import re

import fire

from tj_feeder import HEADERS
from tj_feeder.time_helper import Dates, WorkDay, MONDAY


class Batch:

    def __init__(self):
        self.dates = Dates()

    def create_month_csv_dir(self, directory: str, year: int, month: int) -> None:
        directory = Path(directory) / f'{year}-{month}'
        directory.mkdir(parents=True, exist_ok=True)

        header_line = ','.join(HEADERS)
        workdays = self.dates.get_month_workdays(year=year, month=month)

        for workday in workdays:
            csv_path = directory / f"{workday.strftime('%Y-%m-%d')}.csv"
            if not csv_path.is_file():
                with csv_path.open(mode='w') as file:
                    file.write(f'{header_line}\n')

    def feed_month_csv_dir(self, month_directory: str) -> str:
        # fetching files
        month_directory = Path(month_directory)
        files = [Path(file) for file in glob(f'{month_directory / "*.csv"}')]

        output_str = ''
        last_seen_weekday = MONDAY
        for file in files:
            try:
                work_day = WorkDay(csv_file=file)
            except ValueError as e:
                continue

            # parsing date
            year, month, day = map(int, re.split(r'\D', file.stem))
            current_day = datetime(year, month, day)
            current_weekday = current_day.weekday()

            # building feed strings
            warning_msg = work_day.issue_warnings()

            daily_feed_str = '\n' * (3 if current_weekday < last_seen_weekday else 1)
            if warning_msg:
                daily_feed_str += f'# {warning_msg}\n'
            daily_feed_str += work_day.daily_feed(year=year, month=month, day=day)

            output_str += daily_feed_str

            # updating weekday
            last_seen_weekday = current_weekday

        return output_str


if __name__ == "__main__":
    fire.Fire(Batch)


tj_feeder/tj_feed.py
from typing import Optional

import fire
from loguru import logger

from tj_feeder import configs, time_helper as th


def define(
    starting_hour: Optional[int] = None,
    shift_hours: Optional[int] = None,
    use_minutes: Optional[bool] = None
) -> None:
    """Sets default configuration.

    Args:
        starting_hour (int, optional): Starting hour of the shift. Defaults to None.
        shift_hours (int, optional): How long is the shift in hours. Defaults to None.
        use_minutes (bool, optional): If True, the feed periods will be in minutes; if False, the periods will be in hours. Defaults to None.
    """

    cfg_dict = configs.load()
    if starting_hour is not None:
        print(f'Defining starting hour as {starting_hour}')
        cfg_dict['starting_hour'] = starting_hour

    if shift_hours is not None:
        print(f'Defining shift duration as {shift_hours} hours')
        cfg_dict['shift_hours'] = shift_hours

    if use_minutes is not None:
        print(f'Defining duration unit as {"minutes" if use_minutes else "hours"}')
        cfg_dict['use_minutes'] = use_minutes

    configs.save(cfg_dict)


def feed(
    csv_file: str,
    day: int = 1,
    month: int = 1,
    year: int = 2021,
) -> str:
    """Generates a Daily feed for TaskJuggler from a csv file.

    Args:
        csv_file (str): Path to csv file.
        day (int, optional): Day of entry. Defaults to 1.
        month (int, optional): Month of entry. Defaults to 1.
        year (int, optional): Year of entry. Defaults to 2021.

    Returns:
        str: Daily feed string
    """
    work_day = th.WorkDay(csv_file=csv_file)
    daily_feed_str = work_day.daily_feed(year=year, month=month, day=day)
    work_day.issue_warnings()

    return daily_feed_str


@logger.catch(reraise=True)
def main():
    fire.Fire()


if __name__ == "__main__":
    main()


tj_feeder/time_helper.py
from datetime import timedelta, datetime
from typing import List, Optional, Tuple, Callable
import re

import fire
from loguru import logger
import pandas as pd

from tj_feeder import CFG_FILE, HEADERS, configs


SECONDS_PER_MINUTE = 60
SECONDS_PER_HOUR = SECONDS_PER_MINUTE * 60
MONDAY = 0
SATURDAY = 5
SUNDAY = 6
FLOAT_PATTERN = r"[+]?(\d+(\.\d*)?|\.\d+)"


def td_minutes(td: timedelta) -> int:
    return td.total_seconds() // SECONDS_PER_MINUTE


def td_hours(td: timedelta) -> float:
    return td.total_seconds() / SECONDS_PER_HOUR


def parse_time_string(time: str) -> Tuple[int, float]:
    """Considering the following formats: 30min and 0.5h (a preceding '+' symbol is optional).
    Converts the input from one of those formats to the other.

    Args:
        time (str): Time string

    Raises:
        ValueError: Raised if the input is in neither of the formats

    Returns:
        Tuple[int, float]: Tuple with both formats: (minutes, hours)
    """
    if not re.fullmatch(FLOAT_PATTERN + r'(min|h)', time):
        raise ValueError(f'Invalid period found: {time}. '
                         f'Please inform with either of these formats: '
                         f'XYmin; X.Yh')

    if 'h' in time:
        hours = float(time.replace('h', ''))
        minutes = round(hours * 60)
    else:
        minutes = int(time.replace('+', '').replace('min', ''))
        hours = round(minutes / 60, 2)

    return minutes, hours


class Dates:
    cfg = configs.load()

    def __init__(self):
        try:
            self.holidays_file = self.cfg['holidays_file']
        except KeyError as e:
            raise KeyError('Please set the path to holidays_file with "tj_feed define --holidays-file <path_to_holiday_file>"') from e
        self.holidays = Dates.parse_holidays_file(self.holidays_file)

    @staticmethod
    def parse_holidays_file(holidays_file: str) -> List[Optional[datetime]]:
        """Search for dates in the lines of a file following the format "YYYY-MM-DD" (separator can be any non-digit character).

        Args:
            holidays_file (str): Path to holidays file

        Returns:
            List[Optional[datetime]]: Holidays found in file
        """
        found_dates = []
        with open(holidays_file) as file:
            for line in file:
                match = re.search(r'\d{4}\D\d{2}\D\d{2}', line)
                if match:
                    match_str = match.group(0)
                    formatted = re.sub(r'\D', '-', match_str)
                    found_dates.append(datetime.strptime(formatted, '%Y-%m-%d'))

        return found_dates

    @staticmethod
    def is_weekend(date: datetime) -> bool:
        """Checks if the date is a weekend

        Args:
            date (datetime): Date to check

        Returns:
            bool: Returns True if the day is either saturday or sunday; False, otherwise.
        """
        return date.weekday() in [SATURDAY, SUNDAY]

    @staticmethod
    def is_next_month(date: datetime, month: int) -> bool:
        """Checks if the date is the initial date of the next month.

        Args:
            date (datetime): Date to check
            month (int): Current month

        Returns:
            bool: True if is the date is the first date of the next month. False, otherwise.
        """
        return date.day == Dates.cfg['month_start_workday'] and date.month != month

    def get_list_of_days(self, stop_function: Callable, year: int, month: int, start_workday: int) -> List[Optional[datetime]]:
        """Gets a list of workdays considering the holidays using a function
            to detect the end of the list (non-inclusive).

        Args:
            stop_function (Callable): Function that returns True for the first date not to be included.
            year (int): Current year
            month (int): Current month
            week_start_workday (int): First workday of the period ranging from 1 to 31.

        Returns:
            List[Optional[datetime]]: [description]
        """
        # loop's starting value and increment
        current_dt = datetime(year, month, start_workday)
        day_td = timedelta(days=1)

        workdays = []
        while True:
            # reached end of the week
            if stop_function(current_dt):
                break

            # skipping holidays
            if current_dt in self.holidays \
                or self.is_weekend(current_dt):
                current_dt += day_td
                continue

            # workday
            workdays.append(current_dt)
            current_dt += day_td

        return workdays

    def get_week_workdays(self, year: int, month: int, week_start_workday: int) -> List[Optional[datetime]]:
        """Gets a list of workdays for the week considering the holidays.

        Args:
            year (int): Current year
            month (int): Current month
            week_start_workday (int): First workday of the week. Usually, the date for monday.

        Returns:
            List[Optional[datetime]]: Week's workday dates
        """
        return self.get_list_of_days(
            stop_function=self.is_weekend,
            year=year,
            month=month,
            start_workday=week_start_workday
        )

    def get_month_workdays(self, year: int, month: int) -> List[Optional[datetime]]:
        """Gets a list of workdays for the month considering the holidays.

        Args:
            year (int): Current year
            month (int): Current month

        Returns:
            List[Optional[datetime]]: month's workday dates
        """
        return self.get_list_of_days(
            stop_function=lambda dt: Dates.is_next_month(dt, month),
            year=year,
            month=month,
            start_workday=Dates.cfg['month_start_workday']
        )


class WorkDay:

    def __init__(self, csv_file: str, cfg_file: str = CFG_FILE):
        self.cfg = configs.load(cfg_file=cfg_file)

        self.df = pd.read_csv(csv_file).sort_values(HEADERS[1])

        found_headers = self.df.columns.to_list()
        if found_headers != HEADERS:
            raise ValueError(f'Wrong headers for: "{csv_file}"\n' + \
                             f'Expected headers: {HEADERS}\n' + \
                             f'Found headers: {found_headers}')

        if not len(self.df.values):
            raise ValueError(f'"{csv_file}" is empty.')

        self.minutes_per_day, self.hours_per_day = WorkDay.parse_time_spent(self.df['time_spent'])

        self.worktime_minutes = sum(self.minutes_per_day)
        self.expected_worktime_minutes = self.cfg['shift_hours'] * 60

        self.due_time_td = WorkDay.calculate_due_time(self.expected_worktime_minutes, self.worktime_minutes)
        self.over_time_td = WorkDay.calculate_over_time(self.expected_worktime_minutes, self.worktime_minutes)

    @staticmethod
    def parse_time_spent(time_spent: pd.Series) -> Tuple[List[int], List[int]]:
        minute_hour_times = [parse_time_string(ts) for ts in time_spent]
        minutes_per_day, hours_per_day = tuple(map(list, zip(*minute_hour_times)))
        return minutes_per_day, hours_per_day

    @staticmethod
    def calculate_over_time(expected_worktime_minutes: int, worktime_minutes: int) -> timedelta:
        return timedelta(minutes=max(0, worktime_minutes - expected_worktime_minutes))

    @staticmethod
    def calculate_due_time(expected_worktime_minutes: int, worktime_minutes: int) -> timedelta:
        return timedelta(minutes=min(expected_worktime_minutes - worktime_minutes, expected_worktime_minutes))

    def daily_feed(self, year, month, day):
        daily_feed_str = ''

        # units
        if self.cfg['use_minutes']:
            value_per_day = self.minutes_per_day
            UNIT = 'min'
        else:
            value_per_day = self.hours_per_day
            UNIT = 'h'

        # main loop
        SHIFT_TD = timedelta(hours=self.cfg['shift_hours'])
        cummulative_td = timedelta(minutes=0)
        cur_dt = datetime(year, month, day, self.cfg['starting_hour'])
        for i in range(len(self.minutes_per_day)):
            fmt_time = cur_dt.strftime('%Y-%m-%d-%H:%M')

            # spent time
            spent_time = f'+{value_per_day[i]}{UNIT}'

            # over time
            cummulative_td += timedelta(minutes=self.minutes_per_day[i])
            if cummulative_td > SHIFT_TD:
                spent_time = f'{spent_time:7} {{over_time 1}}'

            # feed line
            daily_feed_str += f"booking {self.df['issue_name'].iloc[i]:30} " + \
                              f"{fmt_time} {spent_time:20} " + \
                              f"# {self.df['issue_description'].iloc[i]}\n"

            # next datetime to show
            cur_dt += timedelta(minutes=self.minutes_per_day[i])

        return daily_feed_str

    def issue_warnings(self) -> str:
        logger.info('Checking missing time and over time...')
        logger.info(f'work_time {self.worktime_minutes:3} minutes')

        warning_msg = ''
        if td_minutes(self.due_time_td):
            warning_msg = f"You are missing {td_hours(self.due_time_td):.2f} hours ({td_minutes(self.due_time_td)} minutes)"
        elif td_minutes(self.over_time_td):
            warning_msg = f"You've worked overtime of {td_hours(self.over_time_td):.2f} hours ({td_minutes(self.over_time_td)} minutes)"

        logger.warning(warning_msg)

        return warning_msg


if __name__ == "__main__":
    fire.Fire()
